text/0000-supporting-code-generation.md

```diff
@@ -0,0 +1,302 @@
+
+# Detailed design
+[design]: #detailed-design
+
+This RFC proposes two changes that will help improve Rust's code generation
+story.
+
+## Source Mapping
+[source mapping]: #source-mapping
+
+Because of the challenges debugging generated code, this RFC proposes that Rust
+be extended to produce and consume a file that contains a mapping from the
+input generated file to the output Rust file.  Lets consider using the rustc
+pretty printer to convert one Rust source into another.  For example, consider
+a simple crate that's made up of two files.  `queen.rs`:
+
+```rust
+pub mod love;
+
+pub struct Person { ... }
+```
+
+and it's submodule, `love.rs`:
+
+```rust
+use super::Person;
+
+pub fn find(people: &[Person]) -> Option<&Person> {
+    people.find(|person| person.lovable())
+}
+```
+
+The pretty printer produces a single output file that merges the two files
+together, and would look something like this:
+
+```
+pub mod love {
+    use super::Person;
+
+    pub fn find(people: &[Person]) -> Option<&Person> {
+        people.find(|person| person.lovable())
+    }
+}
+
+pub struct Person { ... }
+```
+
+By itself, this process loses the information that the module `love`
+came from the file `love.rs`.  To avoid that, the pretty printer will
+instead generate a file, `queen.rs.map`, that conceptually contains the
+following mapping:
+
+| dst line | dst col | source file | src line | src col | token           |
+| -------- | ------- | ----------- | -------- | ------- | --------------  |
+| 0        | 0       | "queen.rs"  | 0        | 0       | pub             |
+| 0        | 4       | "queen.rs"  | 0        | 4       | mod             |
+| 0        | 8       | "queen.rs"  | 0        | 8       | love\_canidates |
+| 0        | 24      | "queen.rs"  | 0        | 24      | ;               |
+| 2        | 0       | "love.rs"   | 0        | 0       | use             |
+
+This mapping will then be used by the Rust compiler during parsing to map
+tokens to their original location.
+
+Rather than Rust developing their own custom mapping file, this RFC proposes
+that Rust adopt the
+[JavaScript Source Map](https://source-map.github.io/)
+[v3 specification](https://docs.google.com/document/d/1U1RGAehQwRypUTovF1KRlpiOFze0b-_2gc6fAH0KY0k/edit)
+This would be done in order to simplify the implementation
+since there are already a number of X-to-JavaScript Source Map generators.
+
+## Source Search Paths
+[paths]: #paths
+
+In order to cut down on the boilerplate necessary including generated source into
+a crate, the Rust Compiler should be extended to support the concept of source
+search paths, similar to GCC's `-I some-path` option, as in
+`rustc -I src -I $OUT_DIR/src`.  When Rust needs to look for some file, it will
+check first in the current directory, then it will iterate through each search
+path until the file is found.
+
+Cargo would then be updated to add the `$OUT_DIR` first in the search path
```

Comment:
Actually, on second thought I think Cargo's gonna need to be a bit smarter here or we need to expand this section a bit more. So let's say there's two project with source trees like:

```
// a
Cargo.toml
lib.rs
foo.rs

// b
Cargo.toml
src/lib.rs
src/foo.rs
```

The "magic build script" for serde will likely mirror the output structure to match the input structure. This means that the build script will generate (assuming no intermediate directory like I mentioned in my comment above this one)

```
// a
$OUT_DIR/foo.rs

// b
$OUT_DIR/src/foo.rs
```

I'd hope that both cases would work, but it's not clear to me how we'd implement this to make them both work. Can we expand on exactly what arguments Cargo is passing in these two situations, along with exactly what the compiler is doing in terms of lookup paths?

Specifically, this section is a little vague to me:

> When Rust needs to look for some file, it will check first in the current directory, then it will iterate through each search path until the file is found.

This to me seems like it can be interpreted as:

* I passed `src/lib.rs` to the compiler
* The compiler saw `mod foo`
* The compiler deduced it needs to look for `src/foo.rs`
* All lookup paths are queried for `src/foo.rs`, where the final lookup path is the current directory

This, however, is incompatible with passing absolute paths to the compiler (which Cargo does frequently). as lookup paths will be queried for an absolute path which fails.